### PR TITLE
feat(fulltext_index): introduce creator

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4246 

## What's changed and what's your intention?

as @coderabbitai

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced a fulltext indexing module with support for configurable analyzers (English and Chinese) and case sensitivity options.
  - Added functionality to create fulltext indexes, push text to the index, and retrieve memory usage during index creation.

- **Bug Fixes**
  - Corrected parameter naming from `capicity` to `capacity` in various functions and tests to ensure consistency and prevent errors.

- **Error Handling**
  - Introduced structured error handling for fulltext indexing with detailed error types and status codes.

### CODE REVIEW

File: src/index/src/fulltext_index/error.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+use common_error::status_code::StatusCode;
+use common_macro::stack_trace_debug;
+use snafu::{Location, Snafu};
+
+#[derive(Snafu)]
+#[snafu(visibility(pub))]
+#[stack_trace_debug]
+pub enum Error {
+    #[snafu(display("IO error"))]
+    IO {
```
Comment (v0y4g3r) [inline src/index/src/fulltext_index/error.rs:28]: Conform with the style: https://github.com/GreptimeTeam/greptimedb/blob/115c74791d123d28ae08aa563e014387b614359c/src/auth/src/error.rs#L34
```suggestion
    Io {
```

File: src/index/src/fulltext_index.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+pub mod create;
+pub mod error;
+
+/// Configuration for fulltext index.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+pub struct Config {
+    /// Analyzer to use for tokenization.
+    pub analyzer: Analyzer,
+
+    /// Whether the index should be case sensitive.
```
Comment (fengjiachun) [inline src/index/src/fulltext_index.rs:26]:
```suggestion
    /// Whether the index should be case-sensitive.
```
This word seems to be usually hyphenated.

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        }
+
+        let tokenizer = builder.build();
+        let tokenizer_manager = TokenizerManager::new();
+        tokenizer_manager.register("default", tokenizer);
+        tokenizer_manager
+    }
+
+    fn sanitize_memory_limit(memory_limit: usize) -> usize {
+        // Port from tantivy::indexer::index_writer::{MEMORY_BUDGET_NUM_BYTES_MIN, MEMORY_BUDGET_NUM_BYTES_MAX}
```
Comment (v0y4g3r) [inline src/index/src/fulltext_index/create/tantivy.rs:86]: Can we just use those consts in tantivy crate?

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        const MEMORY_BUDGET_NUM_BYTES_MAX: usize = u32::MAX as usize - MARGIN_IN_BYTES;
+
+        memory_limit.clamp(MEMORY_BUDGET_NUM_BYTES_MIN, MEMORY_BUDGET_NUM_BYTES_MAX)
+    }
+}
+
+#[async_trait]
+impl FulltextIndexCreator for TantivyFulltextIndexCreator {
+    async fn push_text(&mut self, text: &str) -> Result<()> {
+        if let Some(writer) = &mut self.writer {
```
Comment (v0y4g3r) [inline src/index/src/fulltext_index/create/tantivy.rs:98]: Maybe return an error when writer is already finished?

**Review by v0y4g3r — APPROVED** [review]
Mostly LGTM, index master

**Comment by zhongzc** [comment]
@coderabbitai pause

File: src/index/src/fulltext_index/create/tantivy.rs (hunk 3 above)
Comment (evenyag) [reply src/index/src/fulltext_index/create/tantivy.rs:86]: out-of-topic: Maybe we should limit the total row number of each SST file in compaction and flush.

Comment (zhongzc) [reply src/index/src/fulltext_index/create/tantivy.rs:86]: tanvity does not export these consts

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+                let query_parser = QueryParser::for_index(
+                    &index,
+                    vec![index.schema().get_field(TEXT_FIELD_NAME).unwrap()],
+                );
+                let query = query_parser.parse_query(query).unwrap();
+                let docs = searcher.search(&query, &DocSetCollector).unwrap();
+                let mut res = docs.into_iter().map(|d| d.doc_id).collect::<Vec<_>>();
+                res.sort();
+                assert_eq!(res, expected);
+            }
```
Comment (evenyag) [inline src/index/src/fulltext_index/create/tantivy.rs:169]: nitpick: Is it possible to reuse these codes?

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**